### PR TITLE
catalog: Add durable catalog snapshot type

### PR DESCRIPTION
**Pull request by jkosh44**
This commit adds a durable catalog snapshot type that represents a
complete snapshot of the on-disk state. A trait method was added to
ReadOnlyDurableCatalogState to retrieve the snapshot and the
transaction constructor method was updated to be built from a snapshot.

This achieves a couple of useful things:
- Organizes the code a little bit.
- Puts in some of the foundations for a DurableCatalogState
implementation to maintain an in-memory cache.
- Puts in some of the foundations to allow a transaction to be built
from memory, without going to a durable store.

### Motivation
This PR refactors existing code.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/catalog/src/stash.rs [hunk 1]
```diff
@@ -640,17 +738,27 @@ impl ReadOnlyDurableCatalogState for Connection {
...
+        let cluster_replicas = self.get_cluster_replicas().await?;
+        let introspection_sources = self.get_introspection_source_indexes().await?;
+        let id_allocator = self.get_id_allocators().await?;
+        let configs = self.get_configs().await?;
+        let settings = self.get_settings().await?;
+        let timestamps = self.get_timestamps().await?;
+        let system_object_mappings = self.get_system_items().await?;
+        let system_configurations = self.get_system_configurations().await?;
+        let default_privileges = self.get_default_privileges().await?;
+        let system_privileges = self.get_system_privileges().await?;
```
Comment (jkosh44) [inline src/catalog/src/stash.rs:759]: This is wrong, it needs to be done in a single catalog txn.

File: src/catalog/src/transaction.rs [hunk 2]
```diff
@@ -168,29 +168,90 @@ pub struct Transaction<'a> {
...
+            settings,
+            timestamps,
+            system_object_mappings,
+            system_configurations,
+            default_privileges,
+            system_privileges,
+        }: Snapshot,
     ) -> Result<Transaction, Error> {
+        let databases = databases
+            .into_iter()
```
Comment (mjibson) [inline src/catalog/src/transaction.rs:191]: This looks like each Transaction will go through an additional protobuf decoding then encoding step? Is that right? Is the refactor worth the perf loss?

Comment (jkosh44) [reply src/catalog/src/transaction.rs:191]: Good point, what if we switch the types to use proto directly so we don't have to eat that cost?

**Review by mjibson — APPROVED** [review]
Yeah this looks probably better.

**Pull request merged**